File: problems/2816-double-a-number-represented-as-a-linked-list.ts
export {}

class ListNode {
  val: number
  next: ListNode | null
  constructor(val: number = 0, next: ListNode | null = null) {
    this.val = val
    this.next = next
  }
}

function doubleIt(head: ListNode | null): ListNode | null {
  const stack: ListNode[] = []
  let cur = head
  while (cur) {
    stack.push(cur)
    cur = cur.next
  }

  let carry = 0
  while (stack.length) {
    cur = stack.pop()!
    const val = cur.val * 2 + carry
    cur.val = val % 10
    carry = Math.floor(val / 10)
  }

  return carry === 0 ? cur : new ListNode(carry, cur)
}

const n3 = new ListNode(9)
const n2 = new ListNode(8, n3)
const n1 = new ListNode(1, n2)

doubleIt(n1)

File: problems/2816_double_a_number_represented_as_a_linked_list.py
from typing import Self


class ListNode:
    def __init__(self, val: int = 0, next: Self | None = None):
        self.val = val
        self.next = next


class Solution:
    def doubleIt(self, head: ListNode | None) -> ListNode | None:
        nodes: list[ListNode] = []
        cur = head
        while cur:
            nodes.append(cur)
            cur = cur.next

        carry = 0
        while nodes:
            cur = nodes.pop()
            val = cur.val * 2 + carry
            cur.val = val % 10
            carry = val // 10

        return cur if carry == 0 else ListNode(carry, cur)
